### PR TITLE
Fix queries with hash key and filter

### PR DESCRIPTION
**Pull request by andrewda**
See:
#1240
#1243

### Summary:

Queries using the table hash key and a filter are currently failing due to how we are trying to find the best index. This resolves this issue by performing an additional check to ensure the table index is being used as a last resort, if possible (i.e. if there is an "EQ" operation on the hash key).

#### Schema
```ts
const ExampleModel = dynamoose.model(
  "example", {
    "id": {
      "hashKey": true,
      "type": Number,
    },
    "name": {
      "type": String,
    },
   ...   
  });
```

#### Model
```ts
// Should not fail
const response = await ExampleModel.query("id").eq(111).filter("name").eq("Charlie").exec()
```

### Type (select 1):
- [x] Bug fix
- [ ] Feature implementation
- [ ] Documentation improvement
- [x] Testing improvement

- [ ] Something not listed here

### Is this a breaking change? (select 1):
- [ ] 🚨 YES 🚨
- [x] No
- [ ] I'm not sure

### Is this ready to be merged into Dynamoose? (select 1):
- [x] Yes
- [ ] No

### Are all the tests currently passing on this PR? (select 1):
- [x] Yes
- [ ] No

### CODE REVIEW

**Comment by fishcharlie** [comment]
This looks **really** good at first glance. Wish that @markelrod was able to test this PR, but I'm not sure how to get it working with TypeScript and pointing to a branch. 😢 

@daria-garkusha are you able to review this PR as well?

**Comment by andrewda** [comment]
You should be able to add this to your `package.json` to use this commit:

```
    "dynamoose": "github:andrewda/dynamoose#42767a019c24902984e34387fa903e936b3ae2f1",
```

**Comment by markelrod** [comment]
Let me know when 2.8.1 is ready and I'll test it right away!

File: test/unit/Query.js [hunk 1]
```diff
@@ -800,6 +800,14 @@ describe("Query", () => {
 					});
 				});
 
+				it("Should not throw error if query contains hash key and additional filter expression", () => {
+					queryPromiseResolver = () => ({"Items": []});
+					Model = dynamoose.model("Cat", new dynamoose.Schema({"id": {"type": Number, "hashKey": true}, "name": String}));
+
+					return expect(callType.func(Model.query("id").eq(111).filter("name").eq("Charlie").exec).bind(
+						Model.query("id").eq(111).filter("name").eq("Charlie"))()).to.not.be.rejectedWith("Index can't be found for query.");
```
Comment (fishcharlie) [inline test/unit/Query.js:808]: I think maybe we should change this to be a truthy value as opposed to just verifying that it's not rejected with that specific error. Probably checking to ensure the result is an empty array. That way if it returns a different error this test won't pass.

File: lib/utils/find_best_index.ts [hunk 2]
```diff
@@ -18,5 +23,11 @@ export default function (indexes: ModelIndexes, comparisonChart: ConditionStorag
...
 
 	const index = validIndexes.find((index) => comparisonChart[index._rangeKey]) || validIndexes[0];
-	return index?.IndexName ?? null;
+
+	// If no index is found and hash key exists, use table index
+	if (!index && comparisonChart[tableHashKey]?.type === "EQ") {
+		return {"tableIndex": true};
+	}
+
+	return {"tableIndex": false, "indexName": index?.IndexName ?? null};
```
Comment (fishcharlie) [inline lib/utils/find_best_index.ts:32]: One question. Without this change, is at least 1 of the tests you modified/added failing?

**Review by fishcharlie — COMMENTED** [review]
Looks good. Couple of questions. Would also like to have @daria-garkusha review if possible before merging.

File: test/unit/Query.js (hunk 1 above)
Comment (andrewda) [reply test/unit/Query.js:808]: Agreed, we should check the output and the throw check will be implied.

File: lib/utils/find_best_index.ts (hunk 2 above)
Comment (andrewda) [reply lib/utils/find_best_index.ts:32]: Yep! Anything with `tableHashKey` should fail without this change.

https://github.com/andrewda/dynamoose/blob/42767a019c24902984e34387fa903e936b3ae2f1/test/unit/utils/find_best_index.js#L27-L40

**Comment by andrewda** [comment]
Alright, I've made another update which restructures how we decide to use the table index. First of all, the `ModelIndexes` interface now includes `TableIndex`, which is the index of the table. This exposes the table hash and range keys in `find_best_index`.

1. Our first choice for index is one with both hash key and range key defined.
2. If no index has both hash and range keys defined in the query, then our next choice is the table index (assuming its hash key is present).
3. If we still haven't found an index, then we use the first index with a hash key defined in the query.
4. If still no index, throw an `Index can't be found for query` error.

This also allows us to remove the `can_use_index_of_table` util, since `find_best_index` now also takes the table index into account. This centralizes the logic a little more, making testing more effective and straight forward.

**Comment by andrewda** [comment]
Would definitely love someone who was having the issue previously to test with the latest commit 340ec36f01346e37f5d7b678f91c644f16c5e7af.

```
    "dynamoose": "github:andrewda/dynamoose#340ec36f01346e37f5d7b678f91c644f16c5e7af",
```

**Comment by daria-garkusha** [comment]
@fishcharlie @andrewda I checked this PR, and it works fine for me. The new implementation of `find_best_index` is definitely more precise than before and should cover all cases.

File: lib/Schema.ts [hunk 3]
```diff
@@ -336,11 +336,17 @@ export class Schema {
...
 			AttributeDefinitions,
-			KeySchema,
-			...await this.getIndexes(model)
+			KeySchema
 		};
+
+		const {GlobalSecondaryIndexes, LocalSecondaryIndexes} = await this.getIndexes(model);
+
+		if (GlobalSecondaryIndexes) response.GlobalSecondaryIndexes = GlobalSecondaryIndexes;
+		if (LocalSecondaryIndexes) response.LocalSecondaryIndexes = LocalSecondaryIndexes;
```
Comment (fishcharlie) [inline lib/Schema.ts:347]:
```suggestion
		if (GlobalSecondaryIndexes) {
			response.GlobalSecondaryIndexes = GlobalSecondaryIndexes;
		}
		if (LocalSecondaryIndexes) {
			response.LocalSecondaryIndexes = LocalSecondaryIndexes;
		}
```

**Review by fishcharlie — APPROVED** [review]

**Pull request merged**